controller/src/main/java/org/jboss/as/controller/capability/RuntimeCapability.java

```diff
@@ -329,14 +325,22 @@ public boolean isAllowMultipleRegistrations() {
         assert isDynamicallyNamed();
         assert dynamicElement != null;
         assert dynamicElement.length > 0;
-        return new RuntimeCapability<T>(getName(), serviceValueType, serviceName, runtimeAPI,
+        return new RuntimeCapability<T>(getName(), serviceValueType, getServiceName(), runtimeAPI,
                 getRequirements(), getOptionalRequirements(),
                 getRuntimeOnlyRequirements(), getDynamicRequirements(),
                 getDynamicOptionalRequirements(), allowMultipleRegistrations,
                 dynamicNameMapper, dynamicElement);
 
     }
 
+    private ServiceName getServiceName() {
+        ServiceName result = serviceName;
+        if (result == null && serviceValueType != null) {
+            result = this.serviceName = ServiceNameFactory.parseServiceName(getName());
```

Comment:
Does this need to be thread safe? I just noticed it was volatile is why I ask. I guess though the worst case scenario is we end up with the same `ServiceName`.

Reply:
Exactly. It's volatile so the thread will see the latest at L337 and hopefully avoid falling into the if block, but if there's a race and two threads both see null at L337, it just means we do the work twice.

I could wrap L339 in a synchronized block and check this.serviceName again but I don't really expect that kind of race so I kept it simple for the expected non-race case.